File: code/162.find-peak-element.php
<?php
/*
 * @lc app=leetcode id=162 lang=php
 *
 * [162] Find Peak Element
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function findPeakElement($nums) {
        $length = count($nums);
        if ($nums[0] > $nums[1]) {
            return 0;
        }
        if ($nums[$length - 1] > $nums[$length -1 - 1]) {
            return $length -1;
        }

        $low = 0;
        $high = $length - 1;
        while ($low <= $high) {
            $mid = floor(($low + $high) / 2);

            // mid がpeakの場合
            if ($nums[$mid] > $nums[$mid - 1] && $nums[$mid] > $nums[$mid + 1]) {
                return $mid;
            }

            // 左右にmidより大きい値があるか
            if ($mid != $low && max(array_slice($nums, $low, $mid - $low + 1)) > $nums[$mid]) {
                $high = $mid - 1;
            } elseif ($mid != $high && max(array_slice($nums, $mid, $high - $mid + 1)) > $nums[$mid]) {
                $low = $mid + 1;
            } else {
                break;
            }
        }

        // low == highの場合チェックされないため最後にチェック
        if ($nums[$low] > $nums[$low - 1] && $nums[$low] > $nums[$low + 1]) {
            return $low;
        }

        return -1;
    }
}
// @lc code=end

File: code/162.find-peak-element.go
package code

/*
 * @lc app=leetcode id=162 lang=golang
 *
 * [162] Find Peak Element
 */

// @lc code=start
func findPeakElement(nums []int) int {
	n := len(nums)

	if n == 1 {
		return 0
	}
	// 両端のチェック
	if nums[0] > nums[1] {
		return 0
	}
	if nums[n-1] > nums[n-2] {
		return n - 1
	}

	l := 1
	r := n - 2

	for l <= r {
		m := (l + r) / 2

		if nums[m] > nums[m-1] && nums[m] > nums[m+1] {
			return m
		}

		if nums[m] > nums[m-1] {
			l = m + 1
		} else {
			r = m - 1
		}
	}

	return -1
}

// @lc code=end
